mcfly.fish

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env fish
+
+# Ensure stdin is a tty
+test -t 0; or exit 0
+
+# Avoid loading this file more than once
+if test "$__MCFLY_LOADED" = "loaded"
+  exit 0
+end
+set -g __MCFLY_LOADED "loaded"
+
+# Note: we only use the history file for the session when this file was sourced.
+# Would have to reset this before calling mcfly if you want commands from another session later.
+if not set -q HISTFILE
+  set -gx HISTFILE (set -q XDG_DATA_HOME; and echo $XDG_DATA_HOME; or echo $HOME/.local/share)/fish/(set -q fish_history; and echo $fish_history; or echo fish)_history
+end
+if not test -r "$HISTFILE"
+  echo "McFly: $HISTFILE does not exist or is not readable. Please fix this or set HISTFILE to something else before using McFly." >&2
+  exit 1
+end
+
+# MCFLY_SESSION_ID is used by McFly internally to keep track of the commands from a particular terminal session.
+set -gx MCFLY_SESSION_ID (dd if=/dev/urandom bs=256 count=1 2>/dev/null | env LC_ALL=C tr -dc 'a-zA-Z0-9' | head -c 24)
+
+# Find the binary
+set -q MCFLY_PATH; or set -l MCFLY_PATH (which mcfly)
+# We don't need a MCFLY_HISTORY file because we can get the last command in fish_postexec.
+set -g __MCFLY_CMD $MCFLY_PATH --mcfly_history /dev/null --history_format fish
+
+function __mcfly_save_old_pwd -d 'Save PWD before running command' -e fish_preexec
+  set -g __MCFLY_OLD_PWD "$PWD"
+end
+
+function __mcfly_add_command -d 'Add run commands to McFly database' -e fish_postexec
+  # First, retain return code of last command before we lose it
+  set -l last_status $status
+  # Handle first call of this function after sourcing mcfly.fish, when the old PWD won't be set
+  set -q __MCFLY_OLD_PWD; or set -g __MCFLY_OLD_PWD "$PWD"
+
+  test -n "$MCFLY_DEBUG"; and echo mcfly.fish: Run eval $__MCFLY_CMD add --exit '$last_status' --old-dir '$__MCFLY_OLD_PWD' '$argv[1]'
```

Comment:
This is a distinct difference with fish, yeah - there's less variable expansion / interpretation going on here.  If there are spaces, or special characters used by fish, etc. they're passed through to the command rather than interpreted, and don't need quoting.  I tested commands with spaces, subshells, etc. and they showed up in McFly history as I typed them, not interpreted.

https://fishshell.com/docs/current/tutorial.html#variables  (in particular, the "In bash" line clarifies the difference with respect to spaces.)

Note that the `eval` here is actually removing the quotes, so it's not the quotes saving us, just the difference described above.  (The `eval` is necessary for compatibility with fish 2.x; fish 3 lets you directly use expanded variables like `$__MCFLY_CMD` as commands.)